sdk/go/pulumi/types.go

```diff
@@ -715,7 +720,15 @@ func awaitInputs(ctx context.Context, v, resolved reflect.Value) (bool, bool, []
 	}
 
 	known, secret, deps, err := true, false, make([]Resource, 0), error(nil)
-	switch v.Kind() {
+
+	var shape reflect.Kind
+	if v.Kind() == resolved.Kind() {
```

Comment:
Code below seems to assume pretty strongly that `v.Kind() == resolved.Kind()` as it simultaneously recurs over these. I think this should be safe to not recur when this is not the case, as in our example. It's tricky though..